src/main/java/com/mojang/serialization/codecs/RecordCodecBuilder.java

```diff
@@ -0,0 +1,459 @@
+                        return Stream.of(
+                            function.decoder.keys(ops),
+                            f1.decoder.keys(ops),
+                            f2.decoder.keys(ops),
+                            f3.decoder.keys(ops)
+                        ).flatMap(Function.identity());
+                    }
+
+                    @Override
+                    public String toString() {
+                        return function.decoder + " * " + f1.decoder + " * " + f2.decoder + " * " + f3.decoder;
+                    }
+                }
+            );
+        }
+
+        @Override
+        public <T1, T2, T3, T4, R> App<RecordCodecBuilder.Mu<O>, R> ap4(final App<RecordCodecBuilder.Mu<O>, Function4<T1, T2, T3, T4, R>> func, final App<RecordCodecBuilder.Mu<O>, T1> t1, final App<RecordCodecBuilder.Mu<O>, T2> t2, final App<RecordCodecBuilder.Mu<O>, T3> t3, final App<RecordCodecBuilder.Mu<O>, T4> t4) {
+            final RecordCodecBuilder<O, Function4<T1, T2, T3, T4, R>> function = unbox(func);
+            final RecordCodecBuilder<O, T1> f1 = unbox(t1);
+            final RecordCodecBuilder<O, T2> f2 = unbox(t2);
+            final RecordCodecBuilder<O, T3> f3 = unbox(t3);
+            final RecordCodecBuilder<O, T4> f4 = unbox(t4);
+
+            return new RecordCodecBuilder<>(
+                o -> function.getter.apply(o).apply(
+                    f1.getter.apply(o),
+                    f2.getter.apply(o),
+                    f3.getter.apply(o),
+                    f4.getter.apply(o)
+                ),
+                o -> {
+                    final MapEncoder<Function4<T1, T2, T3, T4, R>> fEncoder = function.encoder.apply(o);
+                    final MapEncoder<T1> e1 = f1.encoder.apply(o);
+                    final T1 v1 = f1.getter.apply(o);
+                    final MapEncoder<T2> e2 = f2.encoder.apply(o);
+                    final T2 v2 = f2.getter.apply(o);
+                    final MapEncoder<T3> e3 = f3.encoder.apply(o);
+                    final T3 v3 = f3.getter.apply(o);
+                    final MapEncoder<T4> e4 = f4.encoder.apply(o);
+                    final T4 v4 = f4.getter.apply(o);
+
+                    return new MapEncoder.Implementation<R>() {
+                        @Override
+                        public <T> RecordBuilder<T> encode(final R input, final DynamicOps<T> ops, final RecordBuilder<T> prefix) {
+                            e1.encode(v1, ops, prefix);
+                            e2.encode(v2, ops, prefix);
+                            e3.encode(v3, ops, prefix);
+                            e4.encode(v4, ops, prefix);
+                            fEncoder.encode((t1, t2, t3, t4) -> input, ops, prefix);
+                            return prefix;
+                        }
+
+                        @Override
+                        public <T> Stream<T> keys(final DynamicOps<T> ops) {
+                            return Stream.of(
```

Comment:
Now that I see it, this is a lot more readable for non-power of 2 number of Streams!